src/recovery/tally.rs

```diff
@@ -290,6 +290,13 @@ impl<I: Iterator<Item = PersistentFragmentLog>> VoteFragmentFilter<I> {
         transaction: &TransactionSlice<P>,
         fragment_id: FragmentId,
     ) -> Result<SpendingCounter, ValidationError> {
+        // check if fragment was processed already
+        if self.replay_protection.contains(&fragment_id) {
+            return Err(ValidationError::DuplicatedFragment { id: fragment_id });
+        }
+
+        self.replay_protection.insert(fragment_id);
```

Comment:
Is early insert OK though? This still may fail, but you already mutated the state.

Reply:
Mmh, other parts of fragment validation are actually stateless (apart from spending counter, but it's not really that important to get it precisely and it's not updated anyway if the transaction is not correct).

Now that I look at it, we allow replay of votes if the initial submission was not inside the voting phase (same for tally), and in general I think you're right, the (real) ledger behavior would be to ignore completely invalid transactions